doc/memos/rdm0.md

```diff
@@ -0,0 +1,148 @@
+license.
+
+## Terminology
+
+This memo uses the following terminology and acronyms:
+
+- RIOT-MC: RIOT Maintainers Community
+- RDM: RIOT Developer Memo
+- PR: git Pull-Request to RIOT's master branch on GitHub 
+- ACK: acknowledgement from a RIOT maintainer, concerning a pull-request to the 
+master branch of RIOT on GitHub
+
+# 1. Introduction
+
+In order to facilitate RIOT maintenance in the long term and at large scale, 
+additional documentation complementing code and usual code documentation is 
+needed. For instance, such memos are expected to describe and give explanations 
+about architectural design decisions, code structure etc. Other memos might
+also describe other aspects of RIOT activity, including, but not limited to,
+RIOT community processes, position with respect to some related external
+technical debates etc.
+
+
+This memo specifies a format for such memos and the process of discussing their 
+content before publication. Format and processes are somewhat inspired by 
+[IETF](http://ietf.org/) processes and documents such as Request For Comments 
+(RFC).
+
+# 2. RIOT Developers Memo Format
+
+RIOT Developer Memos (RDM) are written in English, using the Markdown (.md) 
+format, with less than 80 characters per line. Each memo has a unique 
+identification number indicated in its header, 
+e.g. this memo is RDM0.
+
+The structure of a memo should contain:
+
+1. a preamble as defined below,
+2. the content of the memo structured as needed,
+3. a postamble as defined below.
+
+The preamble of the memo should be structured as such:
+
+1. a header formatted as the header of this memo,
+2. a short abstract,
+3. a status for the memo stating its license and the category of the memo (e.g. 
+that it is a consensus document),
+4. optionally, a terminology section which recalls/disambiguates definition of 
+terms that are key to the memo.
+
+The core content of the memo is structured as deemed necessary. However, during 
+the review process, care must be taken to follow the below principles as much
+as possible:
+
+1. one memo = one problem/aspect,
+2. start by crisply explaining and motivating the one problem/aspect the memo 
+targets,
+3. memos are expected to be of normative and durable nature, hence: keep a memo 
+as concise as possible without impairing clarity, and factor out content that 
+would be likely to be updated too often.
+
+The postamble of the memo should be structured as such:
+
+1. (optional) acknowledgements section thanking contributors other than the 
+authors,
+2. a section with the main references that relate to the content, with a strong
+preference for permanent, self-contained references,
+3. the list of revisions so far and corresponding changelog,
+4. contact information to reach the author(s).
+
+
+# 3. Memo Publishing and Maintenance Process
+
+Each memo is proposed as a pull-request (PR) and discussed the same way RIOT
+code PRs are processed on GitHub.
+
+
```

Comment:
Suggestion: "There is no restriction on when, in the development cycle, a design RDM can be contributed. However, the developer should bear in mind that the design should match the implementation. If a design RDM is submitted after the code PR, and is then contested strongly enough, the implementation may need to be changed."

Reply:
I think there might be a misunderstanding here.
An RDM is not necessarily tied to a specific code contribution PR.
(nevertheless, RDMs are proposed, discussed and merged through a PR).

Reply:
This is to address @kaspar030 's comment asking when in the development process a design-related RDM should be contributed:

https://github.com/RIOT-OS/RIOT/pull/6191#issuecomment-265837501

And to indicate to people that in fact there is no such restriction.

Reply:
@danpetry I see. In my mind, design RDMs would be mostly about 
- documenting models and best practices that are *already* in use, e.g. our driver model; 
- positioning text, e.g. "XXX considered harmful".

In particular we don't want to mandate RDM writing (to avoid TEP-effect in TinyOS).
How about something like: 
"There is no restriction on when, in the development cycle, an RDM can be PRed, but an RDM on design decisions and architecture typically describes what is already (or about to be) merged."

Reply:
Ah, ok, thanks for clarifying.
Strongly agree with not mandating them as per lessons from TinyOS.
I additionally see an important use for them as a platform to foster collaborative discussion and review on new designs. I'd say an RDM on design or architecture should be raised and discussed either before or during the review of a code PR, depending on the nature of the problem. The first if the problem is well-understood and requires a more top-down approach, and the second if it's not and requires a more bottom-up approach where the architecture and the code develop together.
It's not much use discussing a design which has already been implemented, unless we're prepared to change the implementation afterwards.

Reply:
Changed the comment above for clarification.